### PR TITLE
STREAM-40 create delta change summary in script command.

### PR DESCRIPTION
**Pull request by akuma432**

### CODE REVIEW

**Review by satish-sanjeev — APPROVED** [review]

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java [hunk 1]
```diff
@@ -7,13 +7,13 @@
...
 import java.io.File;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.stream.Collectors;
 
 public class ScriptCommand {
   private static final Logger log = LoggerFactory.getLogger(ScriptCommand.class);
+  private static Map<String, Map<String, List>> commonTables;
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:15]: Let us restructure this class to eliminate the requirement for static state. There are two reasons:

1. static state tends to create more static state
2. using the cost in a non-static context, say for a web service, becomes impossible

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java [hunk 2]
```diff
@@ -220,7 +238,111 @@ private static void build(
...
+                entry.getKey(),
+                colDetail.get(Constants.COLUMNS_UPDATED.value()),
+                colDetail.get(Constants.COLUMNS_ADDED.value()),
+                colDetail.get(Constants.COLUMNS_DELETED.value())));
+      }
+    }
+    sb.append("\n");
+    sb.append(
+        String.format(
+            "INCOMPATIBLE TABLES: %s\n",
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:337]: > INCOMPATIBLE TABLES

Let's say TABLES WITH INCOMPATIBLE CHANGES

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java [hunk 3]
```diff
@@ -146,6 +153,17 @@ private static void build(
               });
       writeSchemaMakeFile(
           configuration, configDiff, typeMapping, templateDirectory, outputDirectory);
+
+      try {
+        /*intentionally kept inside try catch block.
+        any exception while generating the delta-change-summary should not block the run. */
+        generateDeltaChangeSummary(configDiff, configuration, typeMapping, outputDirectory);
+      } catch (Exception e) {
+        String errorMsg = "Error generating delta-change-summary file.";
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:162]: > String errorMsg = "Error generating delta-change-summary file.";

Let's change this to:

> String errorMsg = "Error generating delta-change-summary file: " + e;

which will give us the full exception name and error message in the file.

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java [hunk 4]
```diff
@@ -220,7 +238,111 @@ private static void build(
...
 
-  private static TableDefinition getOriginalTable(
+  private static void generateDeltaChangeSummary(
+      ConfigurationDiff configDiff,
+      Configuration config,
+      Map<String, Map<String, String>> typeMapping,
+      String outputDirectory) {
+    // this does not have deleted tables. deleted tables are already removed from this list.
+    List<TableDiff> tableDiffs = configDiff.getTableDiffs();
+    Set<SchemaChanges> validSchemaChanges = ((Jdbc) config.getSource()).getValidSchemaChanges();
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:248]: > (Jdbc) config.getSource()

Feels like we should check if this is true and return void if not as opposed to throwing class cast exception

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java [hunk 5]
```diff
@@ -220,7 +238,111 @@ private static void build(
...
+                tableDiff ->
+                    !tableDiff.allChangesAreCompatible(
+                        JavaHelper.convertJavaMapToScalaMap(typeMapping), validSchemaChanges))
+            .map(incompatibleTable -> incompatibleTable.getDestinationName())
+            .collect(Collectors.toList());
+    addedOrDeletedorIncompatibleTables.put(
+        Constants.TABLES_INCOMPATIBLE.value(), incompatibleTables);
+
+    // now tableDiffs will have updatedTables only.
+    tableDiffs.removeAll(addedTables);
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:273]: tableDiffs is a list, so this is O(n) for each entry in addedTables. These are two moderately sized lists in the worst case, so probaly won't matter.

However as a good practice if we can avoid O(n) operations easily, we should. Could you create a new LinkedHashSet and add TableDiffs list to that at the beginning of the method?

**Review by brockn — CHANGES_REQUESTED** [review]

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java (hunk 1 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java]: Removed all static keywords. Now all are instance method

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java (hunk 2 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java]: updated

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java (hunk 3 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java]: updated

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java (hunk 4 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java]: updated

File: src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java (hunk 5 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/configbuilder/ScriptCommand.java:269]: created a Set at beginning of method.
Set<TableDiff> tableDiffs = new LinkedHashSet<>(configDiff.getTableDiffs());

**Comment by brockn** [comment]
Looks great other than the one logger change.

**Comment by akuma432** [comment]
> Looks great other than the one logger change.

This logger change?  
https://github.com/phdata/streamliner/pull/92/commits/27593655e1185345fb93029a6226ccbbc035437f#diff-c1b47ee919742fb4eb4b5e134f69e3277f1d514e6602a0527a59a6eb7da918eaR15

I removed static keyword from logger as well.

**Comment by brockn** [comment]
Yep I saw that. Maybe my comment on the line of code didn't publish.

static is nuanced. Loggers should almost always be static similar to constants. Other stuff should almost always not be static.

**Review by brockn — APPROVED** [review]

**Pull request merged**